In [1]:
!pip install ipython-sql psycopg2

In [2]:
%load_ext sql

In [3]:
%sql postgresql://postgres:12345@localhost:5432/sql_analysis

In [4]:
%%sql select * from store 
limit 1

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[('CA-2011-103800           ', datetime.date(2013, 1, 3), datetime.date(2013, 1, 7), 'Standard Class', 'DP-13000                 ', 'Darren Powers', 'Consumer', 'United States', 'Houston', 'Texas', 77095, 'Central', 'OFF-PA-10000174', 'Office Supplies', 'Paper', 'Message Book, Wirebound, Four 5 1/2 X 4 Forms/Pg., 200 Dupl. Sets/Book', 16.448, 2, 0.2, 5.5512, 3.2896, 2013, 'old customer', 'Not Returned', 'Not Returned')]

In [ ]:
######################################################################

In [ ]:
##########################################################################################3

# BASIC LEVEL ANALYSIS

In [5]:
#Database Size

%sql SELECT pg_size_pretty(pg_database_size('sql_analysis'));

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[('11 MB',)]

In [6]:
#Table Size

%sql SELECT pg_size_pretty(pg_relation_size('store'));

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[('3456 kB',)]

In [54]:
#row count of data 
%sql SELECT COUNT(*) AS Row_Count FROM store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(9989,)]

In [60]:
#column count of data

In [48]:
%%sql SELECT COUNT(*) AS column_Count FROM information_schema.columns

WHERE table_name = 'store'

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(26,)]

In [61]:
#Check Dataset Information

In [62]:
%%sql SELECT *
FROM INFORMATION_SCHEMA.COLUMNS
WHERE table_name = 'store'

 * postgresql://postgres:***@localhost:5432/sql_analysis
26 rows affected.


[('sql_analysis', 'public', 'store', 'order_date', 3, None, 'YES', 'date', None, None, None, None, None, 0, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'date', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'sales', 18, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'float8', None, None, None, None, '18', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'quantity', 19, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'int4', None, None, None, None, '19', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'discount', 20, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'float8', None, None, None, None, '20', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'profit', 21, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'float8', None, None, None, None, '21', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'discount_amount', 22, None, 'YES', 'double precision', None, None, 53, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'float8', None, None, None, None, '22', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'years', 23, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'int4', None, None, None, None, '23', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'postal_code', 12, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'int4', None, None, None, None, '12', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'ship_date', 4, None, 'YES', 'date', None, None, None, None, None, 0, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'date', None, None, None, None, '4', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'row_id', 1, "nextval('store_row_id_seq'::regclass)", 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'int4', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'product_id', 14, None, 'YES', 'character varying', 75, 300, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'varchar', None, None, None, None, '14', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'category', 15, None, 'YES', 'character varying', 25, 100, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'sql_analysis', 'pg_catalog', 'varchar', None, None, None, None, '15', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('sql_analysis', 'public', 'store', 'sub_category', 16, None, 'YES', 'character varying', 25, 100, None, None, None, None, None, None, None, None, Non

In [ ]:
#get column names of store data 

In [63]:
%%sql select column_name
from INFORMATION_SCHEMA.COLUMNS
where TABLE_NAME='store'


 * postgresql://postgres:***@localhost:5432/sql_analysis
26 rows affected.


[('order_date',),
 ('sales',),
 ('quantity',),
 ('discount',),
 ('profit',),
 ('discount_amount',),
 ('years',),
 ('postal_code',),
 ('ship_date',),
 ('row_id',),
 ('product_id',),
 ('category',),
 ('sub_category',),
 ('product_name',),
 ('customer_duration',),
 ('returned_items',),
 ('return_reason',),
 ('order_id',),
 ('ship_mode',),
 ('customer_id',),
 ('customer_name',),
 ('segment',),
 ('country',),
 ('city',),
 ('states',),
 ('region',)]

In [64]:
#get column names with data type of store data


In [32]:
%%sql select column_name,data_type
from INFORMATION_SCHEMA.COLUMNS
where TABLE_NAME='store'

 * postgresql://postgres:***@localhost:5432/sql_analysis
26 rows affected.


[('order_date', 'date'),
 ('sales', 'double precision'),
 ('quantity', 'integer'),
 ('discount', 'double precision'),
 ('profit', 'double precision'),
 ('discount_amount', 'double precision'),
 ('years', 'integer'),
 ('postal_code', 'integer'),
 ('ship_date', 'date'),
 ('row_id', 'integer'),
 ('product_id', 'character varying'),
 ('category', 'character varying'),
 ('sub_category', 'character varying'),
 ('product_name', 'character varying'),
 ('customer_duration', 'character varying'),
 ('returned_items', 'character varying'),
 ('return_reason', 'character varying'),
 ('order_id', 'character'),
 ('ship_mode', 'character varying'),
 ('customer_id', 'character'),
 ('customer_name', 'character varying'),
 ('segment', 'character varying'),
 ('country', 'character varying'),
 ('city', 'character varying'),
 ('states', 'character varying'),
 ('region', 'character varying')]

In [ ]:
# checking null values of store data

In [67]:
%%sql SELECT * FROM store
WHERE (select column_name
from INFORMATION_SCHEMA.COLUMNS
where TABLE_NAME='store') IS NULL;


 * postgresql://postgres:***@localhost:5432/sql_analysis
(psycopg2.errors.CardinalityViolation) more than one row returned by a subquery used as an expression

[SQL: SELECT * FROM store
WHERE (select column_name
from INFORMATION_SCHEMA.COLUMNS
where TABLE_NAME='store') IS NULL;]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [68]:
%%sql SELECT *
FROM store
WHERE order_date IS NULL OR
sales IS NULL OR
quantity IS NULL OR
discount IS NULL OR
profit IS NULL OR
discount_amount IS NULL OR
years IS NULL OR
postal_code IS NULL OR
ship_date IS NULL OR
row_id IS NULL OR
product_id IS NULL OR
category IS NULL OR
sub_category IS NULL OR
product_name IS NULL OR  
customer_duration IS NULL OR
returned_items IS NULL OR
return_reason IS NULL OR
order_id IS NULL OR
ship_mode IS NULL OR
customer_id IS NULL OR
customer_name IS NULL OR
segment IS NULL OR
country IS NULL OR
city IS NULL OR
states IS NULL OR
region IS NULL 


 * postgresql://postgres:***@localhost:5432/sql_analysis
0 rows affected.


[]

In [ ]:
#Dropping Unnecessary column like Row_ID

In [70]:
%sql ALTER TABLE "store" DROP COLUMN "row_id";

 * postgresql://postgres:***@localhost:5432/sql_analysis
Done.


[]

In [71]:
%sql select * from store limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2011-103800           ', datetime.date(2013, 1, 3), datetime.date(2013, 1, 7), 'Standard Class', 'DP-13000                 ', 'Darren Powers', 'Consumer', 'United States', 'Houston', 'Texas', 77095, 'Central', 'OFF-PA-10000174', 'Office Supplies', 'Paper', 'Message Book, Wirebound, Four 5 1/2 X 4 Forms/Pg., 200 Dupl. Sets/Book', 16.448, 2, 0.2, 5.5512, 3.2896, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-LA-10003223', 'Office Supplies', 'Labels', 'Avery 508', 11.784, 3, 0.2, 4.2717, 2.3568, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-ST-10002743', 'Office Supplies', 'Storage', 'SAFCO Boltless Steel Shelving', 272.736, 3, 0.2, -64.7748, 54.5472, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-BI-10004094', 'Office Supplies', 'Binders', 'GBC Standard Plastic Binding Systems Combs', 3.54, 2, 0.8, -5.487, 2.832, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-141817           ', datetime.date(2013, 1, 5), datetime.date(2013, 1, 12), 'Standard Class', 'MB-18085                 ', 'Mick Brown', 'Consumer', 'United States', 'Philadelphia', 'Pennsylvania', 19143, 'East', 'OFF-AR-10003478', 'Office Supplies', 'Art', 'Avery Hi-Liter EverBold Pen Style Fluorescent Highlighters, 4/Pack', 19.536, 3, 0.2, 4.884, 3.9072, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-130813           ', datetime.date(2013, 1, 6), datetime.date(2013, 1, 8), 'Second Class', 'LS-17230                 ', 'Lycoris Saunders', 'Consumer', 'United States', 'Los Angeles', 'California', 90049, 'West', 'OFF-PA-10002005', 'Office Supplies', 'Paper', 'Xerox 225', 19.44, 3, 0.0, 9.3312, 0.0, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-106054           ', datetime.date(2013, 1, 6), datetime.date(2013, 1, 7), 'First Class', 'JO-15145                 ', "Jack O'Briant", 'Corporate', 'United States', 'Athens', 'Georgia', 30605, 'South', 'OFF-AR-10002399', 'Office Supplies', 'Art', 'Dixon Prang Watercolor Pencils, 10-Color Set with Brush', 12.78, 3, 0.0, 5.2398, 0.0, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-167199           ', datetime.date(2013, 1, 6), datetime.date(2013, 1, 10), 'Standard Class', 'ME-17320                 ', 'Maria Etezadi', 'Home Office', 'United States', 'Henderson', 'Kentucky', 42420, 'South', 'FUR-CH-10004063', 'Furniture', 'Chairs', "Global Deluxe High-Back Manager's Chair", 2573.82, 9, 0.0, 746.4078, 0.0, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-167199           ', datetime.date(2013, 1, 6), datetime.date(2013, 1, 10), 'Standard Class', 'ME-17320                 ', 'Maria Etezadi', 'Home Office', 'United States', 'Henderson', 'Kentucky', 42420, 'South', 'OFF-BI-10004632', 'Office Supplies', 'Binders', 'Ibico Hi-Tech Manual Binding System', 609.98, 2, 0.0, 274.491, 0.0, 2013, 'old customer', 'Not Returned', 'Not Returned'),
 ('CA-2011-167199           ', datetime.date(2013, 1, 6), datetime.date(2013, 1, 10), 'Standard Class', 'ME-17320                 ', 'Maria Etezadi', 'Home Office', 'United States', 'Henderson', 'Kentucky', 42420, 'South', 'OFF-AR-10001662', 'Office Supplies', 'Art', 'Rogers Handheld Barrel Pencil Sharpener', 5.48, 2, 0.0, 1.4796, 0.0, 2013, 'old customer', 'Not Returned', 'Not Returned')]

In [ ]:
#Check the count of United States

In [72]:
%%sql select count(*) AS US_Count 
from store 
where country = 'United States'

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(9989,)]

In [ ]:
#This row isn't important for modeling purposes, but important for auto-generating latitude and longitude on Tableau. So, We won't drop it.


# PRODUCT LEVEL ANALYSIS

In [ ]:
#What are the unique product categories?

In [73]:
%sql select distinct (Category) from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
3 rows affected.


[('Furniture',), ('Office Supplies',), ('Technology',)]

In [ ]:
#What is the number of products in each category?

In [74]:
%%sql SELECT Category, count(*) AS No_of_Products
FROM store
GROUP BY Category
order by  count(*) desc

 * postgresql://postgres:***@localhost:5432/sql_analysis
3 rows affected.


[('Office Supplies', 6021), ('Furniture', 2121), ('Technology', 1847)]

In [ ]:
#Find the number of Subcategories products that are divided.

In [75]:
%%sql select count(distinct (Sub_Category)) As No_of_Sub_Categories
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(17,)]

In [ ]:
#Find the number of products in each sub-category.

In [76]:
%%sql SELECT Sub_Category, count(*) As No_of_products
FROM store
GROUP BY Sub_Category
order by  count(*) desc

 * postgresql://postgres:***@localhost:5432/sql_analysis
17 rows affected.


[('Binders', 1522),
 ('Paper', 1370),
 ('Furnishings', 957),
 ('Phones', 889),
 ('Storage', 846),
 ('Art', 796),
 ('Accessories', 775),
 ('Chairs', 617),
 ('Appliances', 462),
 ('Labels', 364),
 ('Tables', 319),
 ('Envelopes', 254),
 ('Bookcases', 228),
 ('Fasteners', 217),
 ('Supplies', 190),
 ('Machines', 115),
 ('Copiers', 68)]

In [ ]:
#Find the number of unique product names.

In [77]:
%%sql select count(distinct (Product_Name)) As No_of_unique_products
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(1849,)]

In [ ]:
#Which are the Top 10 Products that are ordered frequently?

In [78]:
%%sql SELECT Product_Name, count(*) AS No_of_products
FROM store
GROUP BY Product_Name
order by  count(*) desc
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('Staple envelope', 48),
 ('Easy-staple paper', 46),
 ('Staples', 46),
 ('Avery Non-Stick Binders', 20),
 ('Staples in misc. colors', 19),
 ('Staple remover', 18),
 ('KI Adjustable-Height Table', 18),
 ('Storex Dura Pro Binders', 17),
 ('Staple-based wall hangings', 16),
 ('Logitech 910-002974 M325 Wireless Mouse for Web Scrolling', 15)]

In [ ]:
#Calculate the cost for each Order_Id with respective Product Name.

In [5]:
%%sql select Order_Id,Product_Name,ROUND(CAST((sales-profit) AS NUMERIC), 2)as cost
from store
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2011-103800           ', 'Message Book, Wirebound, Four 5 1/2 X 4 Forms/Pg., 200 Dupl. Sets/Book', Decimal('10.90')),
 ('CA-2011-112326           ', 'Avery 508', Decimal('7.51')),
 ('CA-2011-112326           ', 'SAFCO Boltless Steel Shelving', Decimal('337.51')),
 ('CA-2011-112326           ', 'GBC Standard Plastic Binding Systems Combs', Decimal('9.03')),
 ('CA-2011-141817           ', 'Avery Hi-Liter EverBold Pen Style Fluorescent Highlighters, 4/Pack', Decimal('14.65')),
 ('CA-2011-130813           ', 'Xerox 225', Decimal('10.11')),
 ('CA-2011-106054           ', 'Dixon Prang Watercolor Pencils, 10-Color Set with Brush', Decimal('7.54')),
 ('CA-2011-167199           ', "Global Deluxe High-Back Manager's Chair", Decimal('1827.41')),
 ('CA-2011-167199           ', 'Ibico Hi-Tech Manual Binding System', Decimal('335.49')),
 ('CA-2011-167199           ', 'Rogers Handheld Barrel Pencil Sharpener', Decimal('4.00'))]

In [ ]:
#Calculate % profit for each Order_ID with respective Product Name.

In [6]:
%%sql select Order_Id,Product_Name,ROUND(CAST((profit/((sales-profit))*100)AS NUMERIC),2) as percentage_profit 
from store
limit 10



 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2011-103800           ', 'Message Book, Wirebound, Four 5 1/2 X 4 Forms/Pg., 200 Dupl. Sets/Book', Decimal('50.94')),
 ('CA-2011-112326           ', 'Avery 508', Decimal('56.86')),
 ('CA-2011-112326           ', 'SAFCO Boltless Steel Shelving', Decimal('-19.19')),
 ('CA-2011-112326           ', 'GBC Standard Plastic Binding Systems Combs', Decimal('-60.78')),
 ('CA-2011-141817           ', 'Avery Hi-Liter EverBold Pen Style Fluorescent Highlighters, 4/Pack', Decimal('33.33')),
 ('CA-2011-130813           ', 'Xerox 225', Decimal('92.31')),
 ('CA-2011-106054           ', 'Dixon Prang Watercolor Pencils, 10-Color Set with Brush', Decimal('69.49')),
 ('CA-2011-167199           ', "Global Deluxe High-Back Manager's Chair", Decimal('40.85')),
 ('CA-2011-167199           ', 'Ibico Hi-Tech Manual Binding System', Decimal('81.82')),
 ('CA-2011-167199           ', 'Rogers Handheld Barrel Pencil Sharpener', Decimal('36.99'))]

#Calculate percentage profit and group by them with Product Name and Order_Id.
#Introducing method using WITH 

In [87]:
%%sql WITH store_new as(
select a.*,b.percentage_profit
from store as a
left join
(select ((profit/((sales-profit))*100)) as percentage_profit,order_id,Product_Name from store
group by percentage_profit,Product_Name,order_id) as b
on a.order_id=b.order_id)
select * from store_new
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2011-103800           ', datetime.date(2013, 1, 3), datetime.date(2013, 1, 7), 'Standard Class', 'DP-13000                 ', 'Darren Powers', 'Consumer', 'United States', 'Houston', 'Texas', 77095, 'Central', 'OFF-PA-10000174', 'Office Supplies', 'Paper', 'Message Book, Wirebound, Four 5 1/2 X 4 Forms/Pg., 200 Dupl. Sets/Book', 16.448, 2, 0.2, 5.5512, 3.2896, 2013, 'old customer', 'Not Returned', 'Not Returned', 50.94339622641508),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-LA-10003223', 'Office Supplies', 'Labels', 'Avery 508', 11.784, 3, 0.2, 4.2717, 2.3568, 2013, 'old customer', 'Not Returned', 'Not Returned', 56.86274509803921),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-LA-10003223', 'Office Supplies', 'Labels', 'Avery 508', 11.784, 3, 0.2, 4.2717, 2.3568, 2013, 'old customer', 'Not Returned', 'Not Returned', -60.78431372549019),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-LA-10003223', 'Office Supplies', 'Labels', 'Avery 508', 11.784, 3, 0.2, 4.2717, 2.3568, 2013, 'old customer', 'Not Returned', 'Not Returned', -19.19191919191919),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-ST-10002743', 'Office Supplies', 'Storage', 'SAFCO Boltless Steel Shelving', 272.736, 3, 0.2, -64.7748, 54.5472, 2013, 'old customer', 'Not Returned', 'Not Returned', 56.86274509803921),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-ST-10002743', 'Office Supplies', 'Storage', 'SAFCO Boltless Steel Shelving', 272.736, 3, 0.2, -64.7748, 54.5472, 2013, 'old customer', 'Not Returned', 'Not Returned', -60.78431372549019),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-ST-10002743', 'Office Supplies', 'Storage', 'SAFCO Boltless Steel Shelving', 272.736, 3, 0.2, -64.7748, 54.5472, 2013, 'old customer', 'Not Returned', 'Not Returned', -19.19191919191919),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-BI-10004094', 'Office Supplies', 'Binders', 'GBC Standard Plastic Binding Systems Combs', 3.54, 2, 0.8, -5.487, 2.832, 2013, 'old customer', 'Not Returned', 'Not Returned', 56.86274509803921),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-BI-10004094', 'Office Supplies', 'Binders', 'GBC Standard Plastic Binding Systems Combs', 3.54, 2, 0.8, -5.487, 2.832, 2013, 'old customer', 'Not Returned', 'Not Returned', -60.78431372549019),
 ('CA-2011-112326           ', datetime.date(2013, 1, 4), datetime.date(2013, 1, 8), 'Standard Class', 'PO-19195                 ', 'Phillina Ober', 'Home Office', 'United States', 'Naperville', 'Illinois', 60540, 'Central', 'OFF-BI-10004094', 'Office Supplies', 'Binders', 'GBC 

In [ ]:
#Same Thing Using normal method without creating any temporary data. Here, This can be only viewed for one time and we can't merge with the current dataset in this process.

In [88]:
%%sql select  product_id,order_id,((profit/((sales-profit))*100)) as percentage_profit
from store
group by product_id,order_id,percentage_profit
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('OFF-BI-10002609', 'US-2013-110156           ', -62.962962962962976),
 ('OFF-AR-10002257', 'CA-2011-101427           ', 15.942028985507248),
 ('TEC-PH-10000455', 'CA-2011-124688           ', 40.8450704225352),
 ('FUR-CH-10003199', 'US-2013-134488           ', -7.894736842105263),
 ('OFF-PA-10001947', 'CA-2012-163104           ', 96.078431372549),
 ('OFF-AR-10003478', 'CA-2011-155208           ', 33.33333333333333),
 ('FUR-CH-10003774', 'US-2013-127971           ', -19.54022988505747),
 ('TEC-AC-10004396', 'CA-2012-146675           ', -10.112359550561797),
 ('FUR-CH-10004754', 'CA-2013-118899           ', 9.589041095890412),
 ('OFF-ST-10001590', 'US-2012-159982           ', 8.108108108108107)]

/* Where can we trim some loses? 
   In Which products?
   We can do this by calculating the average sales and profits, and comparing the values to that average.
   If the sales or profits are below average, then they are not best sellers and 
   can be analyzed deeper to see if its worth selling thema anymore. */

In [89]:
%%sql SELECT round(cast(AVG(sales) as numeric),2) AS avg_sales
FROM store;

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(Decimal('229.83'),)]

#the average sales on any given product is 229.8, so approx. 230.

In [90]:
%%sql SELECT round(cast(AVG(Profit)as numeric),2) AS avg_profit
FROM store;

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(Decimal('28.65'),)]


-- the average profit on any given product is 28.6, or approx 29.

In [ ]:
#Average sales per sub-cat

In [91]:
%%sql SELECT round(cast(AVG(sales) as numeric),2) AS avg_sales, Sub_Category
FROM store
GROUP BY Sub_Category
ORDER BY avg_sales asc
limit 9;


 * postgresql://postgres:***@localhost:5432/sql_analysis
9 rows affected.


[(Decimal('13.94'), 'Fasteners'),
 (Decimal('34.07'), 'Art'),
 (Decimal('34.30'), 'Labels'),
 (Decimal('57.28'), 'Paper'),
 (Decimal('64.87'), 'Envelopes'),
 (Decimal('95.83'), 'Furnishings'),
 (Decimal('133.64'), 'Binders'),
 (Decimal('215.97'), 'Accessories'),
 (Decimal('229.77'), 'Appliances')]

--The sales of these Sub_category products are below the average sales.

In [ ]:
#Average profit per sub-cat

In [92]:
%%sql SELECT round(cast(AVG(Profit)as numeric),2) AS avg_prof,Sub_Category
FROM store
GROUP BY Sub_Category
ORDER BY avg_prof asc
limit 11;

 * postgresql://postgres:***@localhost:5432/sql_analysis
11 rows affected.


[(Decimal('-55.57'), 'Tables'),
 (Decimal('-15.23'), 'Bookcases'),
 (Decimal('-6.26'), 'Supplies'),
 (Decimal('4.38'), 'Fasteners'),
 (Decimal('8.20'), 'Art'),
 (Decimal('13.65'), 'Furnishings'),
 (Decimal('15.24'), 'Labels'),
 (Decimal('19.85'), 'Binders'),
 (Decimal('24.86'), 'Paper'),
 (Decimal('25.15'), 'Storage'),
 (Decimal('27.42'), 'Envelopes')]

--The profit of these Sub_category products are below the average profit.
-- "Minus sign" Respresnts that those products are in losses.

# CUSTOMER LEVEL ANALYSIS

In [ ]:
#What is the number of unique customer IDs?

In [93]:
%%sql select count(distinct (Customer_id)) as no_of_unique_custd_ID
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(793,)]

In [ ]:
#Find those customers who registered during 2014-2016.

In [13]:
%%sql select distinct  (Customer_ID), Order_ID,city, Postal_Code
from store
where Customer_Id is not null
limit 10


 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('NP-18685                 ', 'CA-2012-141768           ', 'San Francisco', 94109),
 ('VB-21745                 ', 'CA-2014-155698           ', 'Columbus', 31907),
 ('DH-13075                 ', 'CA-2012-159534           ', 'New York City', 10035),
 ('MD-17860                 ', 'CA-2014-153654           ', 'Asheville', 28806),
 ('JH-16180                 ', 'CA-2012-126739           ', 'New York City', 10035),
 ('NZ-18565                 ', 'CA-2013-110772           ', 'Columbus', 43229),
 ('BD-11620                 ', 'CA-2014-130211           ', 'Lawton', 73505),
 ('BD-11320                 ', 'CA-2012-121699           ', 'Detroit', 48227),
 ('SP-20920                 ', 'CA-2013-105291           ', 'San Luis Obispo', 93405),
 ('NP-18670                 ', 'CA-2014-150091           ', 'Lakewood', 8701)]

In [ ]:
#Calculate Total Frequency of each order id by each customer Name in descending order.

In [10]:
%%sql select order_id, customer_id, count(Order_Id) as total_order_id
from store
group by order_id,customer_id
order by total_order_id desc
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2014-100111           ', 'SV-20365                 ', 14),
 ('CA-2014-157987           ', 'AC-10615                 ', 12),
 ('CA-2013-165330           ', 'WB-21850                 ', 11),
 ('US-2013-108504           ', 'PP-18955                 ', 11),
 ('US-2012-126977           ', 'PF-19120                 ', 10),
 ('CA-2013-105732           ', 'AG-10270                 ', 10),
 ('CA-2012-131338           ', 'NP-18325                 ', 10),
 ('CA-2011-106439           ', 'GG-14650                 ', 9),
 ('CA-2012-164882           ', 'SG-20080                 ', 9),
 ('US-2012-163433           ', 'MP-17965                 ', 9)]

In [ ]:
#Calculate  cost of each customer name.

In [12]:
%%sql select order_id, customer_id, City, Quantity,sales,(sales-profit) as costs,profit
from store
group by order_id,customer_id,City,Quantity,Costs,sales,profit
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('CA-2014-121293           ', 'JC-15775                 ', 'Philadelphia', 4, 20.736, 13.4784, 7.2576),
 ('CA-2011-127866           ', 'JD-15895                 ', 'Riverside', 4, 13.12, 9.315199999999999, 3.8048),
 ('US-2012-126753           ', 'SP-20860                 ', 'Philadelphia', 4, 519.792, 632.4136000000001, -112.6216),
 ('CA-2014-143112           ', 'TS-21370                 ', 'New York City', 4, 87.8, 43.9, 43.9),
 ('CA-2013-161907           ', 'JH-16180                 ', 'Philadelphia', 4, 769.184, 932.6356, -163.4516),
 ('CA-2014-113355           ', 'SJ-20215                 ', 'Grand Prairie', 3, 317.058, 335.1756, -18.1176),
 ('CA-2013-123274           ', 'GT-14710                 ', 'New York City', 3, 242.94, 233.2224, 9.7176),
 ('CA-2012-167010           ', 'VT-21700                 ', 'Philadelphia', 4, 10.428, 17.380000000000003, -6.952),
 ('CA-2014-127432           ', 'AD-10180                 ', 'Great Falls', 3, 1126.02, 1069.719, 56.301),
 ('CA-2012-159534           ', 'DH-13075                 ', 'New York City', 2, 199.98, 145.9854, 53.9946)]

In [ ]:
#Display No of Customers in each region in descending order.

In [97]:
%%sql select Region, count(*) as No_of_Customers
from store
group by region
order by no_of_customers desc

 * postgresql://postgres:***@localhost:5432/sql_analysis
4 rows affected.


[('West', 3202), ('East', 2847), ('Central', 2321), ('South', 1619)]

In [ ]:
#Find Top 10 customers who order frequently.

In [98]:
%%sql SELECT Customer_Name, count(*) as no_of_order
FROM store
GROUP BY Customer_Name
order by  count(*) desc
limit 10

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('William Brown', 37),
 ('Paul Prost', 34),
 ('Matt Abelman', 34),
 ('John Lee', 34),
 ('Chloris Kastensmidt', 32),
 ('Seth Vernon', 32),
 ('Jonathan Doherty', 32),
 ('Edward Hooks', 32),
 ('Zuschuss Carroll', 31),
 ('Arthur Prichep', 31)]

In [ ]:
#Find Top 20 Customers who benefitted the store.

In [99]:
%%sql SELECT Customer_Name, Profit, City, States
FROM store
GROUP BY Customer_Name,Profit,City,States
order by  Profit desc
limit 20

 * postgresql://postgres:***@localhost:5432/sql_analysis
20 rows affected.


[('Tamara Chand', 8399.976, 'Lafayette', 'Indiana'),
 ('Raymond Buch', 6719.9808, 'Seattle', 'Washington'),
 ('Hunter Lopez', 5039.9856, 'Newark', 'Delaware'),
 ('Adrian Barton', 4946.37, 'Detroit', 'Michigan'),
 ('Sanjit Chand', 4630.4755, 'Minneapolis', 'Minnesota'),
 ('Tom Ashbrook', 3919.9888, 'New York City', 'New York'),
 ('Christopher Martinez', 3177.475, 'Atlanta', 'Georgia'),
 ('Sanjit Engle', 2799.984, 'Arlington', 'Virginia'),
 ('Daniel Raglin', 2591.9568, 'Providence', 'Rhode Island'),
 ('Andy Reiter', 2504.2216, 'Jackson', 'Michigan'),
 ('Karen Daniels', 2400.9657, 'Yonkers', 'New York'),
 ('Bill Shonely', 2365.9818, 'Lakewood', 'New Jersey'),
 ('Harry Marie', 2302.9671, 'Springfield', 'Missouri'),
 ('Tom Boeckenhauer', 2239.9872, 'New York City', 'New York'),
 ('Keith Dawkins', 2229.024, 'New York City', 'New York'),
 ('Nathan Mautz', 1995.99, 'New York City', 'New York'),
 ('Jane Waco', 1906.485, 'Sacramento', 'California'),
 ('Shirley Daniels', 1906.485, 'Midland', 'Michigan'),
 ('Steven Roelle', 1668.205, 'New York City', 'New York'),
 ('Dennis Pardue', 1644.2913, 'North Las Vegas', 'Nevada')]

In [ ]:
#Which state(s) is the superstore most succesful in? Least?
#Top 10 results:

In [100]:
%%sql SELECT round(cast(SUM(sales) as numeric),2) AS state_sales, States
FROM Store
GROUP BY States
ORDER BY state_sales DESC
OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[(Decimal('310876.27'), 'New York'),
 (Decimal('170188.05'), 'Texas'),
 (Decimal('138641.27'), 'Washington'),
 (Decimal('116511.91'), 'Pennsylvania'),
 (Decimal('89086.57'), 'Florida'),
 (Decimal('80166.10'), 'Illinois'),
 (Decimal('77871.00'), 'Ohio'),
 (Decimal('76269.61'), 'Michigan'),
 (Decimal('70636.72'), 'Virginia'),
 (Decimal('55603.16'), 'North Carolina')]

# ORDER LEVEL ANALYSIS 

In [ ]:
#number of unique orders

In [101]:
%%sql select count(distinct (Order_ID)) as no_of_unique_orders
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(5007,)]

In [ ]:
#Find Sum Total Sales of Superstore.

In [102]:
%%sql select round(cast(SUM(sales) as numeric),2) as Total_Sales
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(Decimal('2295810.15'),)]

In [ ]:
#Calculate the time taken for an order to ship and converting the no. of days in int format.

In [103]:
%%sql select order_id,customer_id,customer_name,city,states, (ship_date-order_date) as delivery_duration
from store
order by delivery_duration desc
limit 20

 * postgresql://postgres:***@localhost:5432/sql_analysis
20 rows affected.


[('CA-2011-148586           ', 'AZ-10750                 ', 'Annie Zypern', 'New York City', 'New York', 7),
 ('CA-2011-147298           ', 'AG-10300                 ', 'Aleksandra Gannaway', 'Los Angeles', 'California', 7),
 ('CA-2011-118339           ', 'BN-11515                 ', 'Bradley Nguyen', 'Lakeville', 'Minnesota', 7),
 ('CA-2011-118339           ', 'BN-11515                 ', 'Bradley Nguyen', 'Lakeville', 'Minnesota', 7),
 ('CA-2011-110555           ', 'MM-18055                 ', 'Michelle Moray', 'Great Falls', 'Montana', 7),
 ('CA-2011-147298           ', 'AG-10300                 ', 'Aleksandra Gannaway', 'Los Angeles', 'California', 7),
 ('CA-2011-154165           ', 'DL-13315                 ', 'Delfina Latchford', 'Chicago', 'Illinois', 7),
 ('CA-2011-118339           ', 'BN-11515                 ', 'Bradley Nguyen', 'Lakeville', 'Minnesota', 7),
 ('CA-2011-118339           ', 'BN-11515                 ', 'Bradley Nguyen', 'Lakeville', 'Minnesota', 7),
 ('CA-2011-118339           ', 'BN-11515                 ', 'Bradley Nguyen', 'Lakeville', 'Minnesota', 7),
 ('CA-2011-100363           ', 'JM-15655                 ', 'Jim Mitchum', 'Glendale', 'Arizona', 7),
 ('CA-2011-100363           ', 'JM-15655                 ', 'Jim Mitchum', 'Glendale', 'Arizona', 7),
 ('CA-2011-110555           ', 'MM-18055                 ', 'Michelle Moray', 'Great Falls', 'Montana', 7),
 ('CA-2011-110555           ', 'MM-18055                 ', 'Michelle Moray', 'Great Falls', 'Montana', 7),
 ('CA-2011-123400           ', 'EB-13930                 ', 'Eric Barreto', 'Saint Petersburg', 'Florida', 7),
 ('CA-2011-123400           ', 'EB-13930                 ', 'Eric Barreto', 'Saint Petersburg', 'Florida', 7),
 ('US-2011-140116           ', 'KT-16480                 ', 'Kean Thornton', 'Denver', 'Colorado', 7),
 ('US-2011-140116           ', 'KT-16480                 ', 'Kean Thornton', 'Denver', 'Colorado', 7),
 ('CA-2011-141817           ', 'MB-18085                 ', 'Mick Brown', 'Philadelphia', 'Pennsylvania', 7),
 ('CA-2011-135755           ', 'NC-18415                 ', 'Nathan Cano', 'Yonkers', 'New York', 7)]

In [ ]:
#Extract the year  for respective order ID and Customer ID with quantity.

In [104]:
%%sql select order_id,customer_id,quantity,EXTRACT(YEAR from Order_Date) 
from store
group by order_id,customer_id,quantity,EXTRACT(YEAR from Order_Date) 
order by quantity desc

 * postgresql://postgres:***@localhost:5432/sql_analysis
8614 rows affected.


[('CA-2011-154165           ', 'DL-13315                 ', 14, Decimal('2013')),
 ('CA-2012-120621           ', 'JW-16075                 ', 14, Decimal('2014')),
 ('CA-2011-158337           ', 'KA-16525                 ', 14, Decimal('2013')),
 ('CA-2014-123491           ', 'JK-15205                 ', 14, Decimal('2016')),
 ('US-2013-103674           ', 'AP-10720                 ', 14, Decimal('2015')),
 ('CA-2013-105732           ', 'AG-10270                 ', 14, Decimal('2015')),
 ('CA-2014-130036           ', 'BP-11185                 ', 14, Decimal('2016')),
 ('CA-2012-103135           ', 'SS-20515                 ', 14, Decimal('2014')),
 ('CA-2011-120768           ', 'IM-15070                 ', 14, Decimal('2013')),
 ('CA-2014-169859           ', 'MP-18175                 ', 14, Decimal('2016')),
 ('CA-2014-161410           ', 'CC-12220                 ', 14, Decimal('2016')),
 ('CA-2011-115259           ', 'RC-19960                 ', 14, Decimal('2013')),
 ('CA-2013-116596           ', 'BW-11200                 ', 14, Decimal('2015')),
 ('US-2012-119312           ', 'CS-12400                 ', 14, Decimal('2014')),
 ('CA-2013-141887           ', 'MP-17470                 ', 14, Decimal('2015')),
 ('CA-2013-140571           ', 'SJ-20125                 ', 14, Decimal('2015')),
 ('CA-2011-154599           ', 'KN-16450                 ', 14, Decimal('2013')),
 ('CA-2012-146563           ', 'CB-12025                 ', 14, Decimal('2014')),
 ('US-2014-155425           ', 'AB-10600                 ', 14, Decimal('2016')),
 ('CA-2014-151750           ', 'JM-15250                 ', 14, Decimal('2016')),
 ('US-2012-164448           ', 'DK-12835                 ', 14, Decimal('2014')),
 ('CA-2011-163447           ', 'TB-21190                 ', 14, Decimal('2013')),
 ('CA-2012-149713           ', 'TG-21640                 ', 14, Decimal('2014')),
 ('CA-2014-164028           ', 'JL-15835                 ', 14, Decimal('2016')),
 ('CA-2013-142405           ', 'SE-20110                 ', 14, Decimal('2015')),
 ('CA-2012-104241           ', 'AG-10495                 ', 14, Decimal('2014')),
 ('CA-2013-145583           ', 'LC-16885                 ', 14, Decimal('2015')),
 ('CA-2014-152702           ', 'SN-20710                 ', 14, Decimal('2016')),
 ('CA-2012-135489           ', 'GW-14605                 ', 14, Decimal('2014')),
 ('US-2011-151015           ', 'BD-11500                 ', 13, Decimal('2013')),
 ('CA-2014-132234           ', 'MY-17380                 ', 13, Decimal('2016')),
 ('CA-2013-117121           ', 'AB-10105                 ', 13, Decimal('2015')),
 ('CA-2012-116638           ', 'JH-15985                 ', 13, Decimal('2014')),
 ('CA-2014-117457           ', 'KH-16510                 ', 13, Decimal('2016')),
 ('CA-2014-147942           ', 'MS-17365                 ', 13, Decimal('2016')),
 ('US-2014-147669           ', 'SV-20935                 ', 13, Decimal('2016')),
 ('CA-2011-134215           ', 'ML-17395                 ', 13, Decimal('2013')),
 ('CA-2012-148712           ', 'JM-15535                 ', 13, Decimal('2014')),
 ('CA-2014-129378           ', 'NS-18505                 ', 13, Decimal('2016')),
 ('US-2011-155502           ', 'SD-20485                 ', 13, Decimal('2013')),
 ('CA-2011-101364           ', 'TW-21025                 ', 13, Decimal('2013')),
 ('US-2012-126214           ', 'JS-15880                 ', 13, Decimal('2014')),
 ('CA-2011-114321           ', 'NC-18535                 ', 13, Decimal('2013')),
 ('CA-2014-114524           ', 'EG-13900                 ', 13, Decimal('2016')),
 ('US-2013-165505           ', 'CB-12535                 ', 13, Decimal('2015')),
 ('CA-2011-106439           ', 'GG-14650                 ', 13, Decimal('2013')),
 ('CA-2014-129021           ', 'PO-18850                 ', 13, Decimal('2016')),
 ('CA-2012-138898           ', 'JH-16180                 ', 13, Decimal('2014')),
 ('CA-2011-153927           ', 'LL-16840                 ', 13, 

In [ ]:
#What is the Sales impact?

In [105]:
%%sql SELECT EXTRACT(YEAR from Order_Date), Sales, round(cast(((profit/((sales-profit))*100))as numeric),2) as profit_percentage
FROM store
GROUP BY EXTRACT(YEAR from Order_Date), Sales, profit_percentage
order by  profit_percentage 
limit 20

 * postgresql://postgres:***@localhost:5432/sql_analysis
20 rows affected.


[(Decimal('2014'), 143.128, Decimal('-73.33')),
 (Decimal('2014'), 24.588, Decimal('-73.33')),
 (Decimal('2016'), 1.624, Decimal('-73.33')),
 (Decimal('2016'), 2.688, Decimal('-73.33')),
 (Decimal('2016'), 1.556, Decimal('-72.97')),
 (Decimal('2016'), 66.284, Decimal('-72.97')),
 (Decimal('2016'), 1.392, Decimal('-72.97')),
 (Decimal('2015'), 93.032, Decimal('-72.97')),
 (Decimal('2015'), 26.406, Decimal('-72.97')),
 (Decimal('2015'), 48.784, Decimal('-72.97')),
 (Decimal('2016'), 73.176, Decimal('-72.97')),
 (Decimal('2014'), 70.97, Decimal('-72.97')),
 (Decimal('2014'), 19.568, Decimal('-72.97')),
 (Decimal('2016'), 33.62, Decimal('-72.97')),
 (Decimal('2015'), 2.334, Decimal('-72.97')),
 (Decimal('2016'), 5.588, Decimal('-72.97')),
 (Decimal('2013'), 3.16, Decimal('-72.97')),
 (Decimal('2015'), 4.356, Decimal('-72.97')),
 (Decimal('2015'), 62.79, Decimal('-72.60')),
 (Decimal('2016'), 9.324, Decimal('-72.60'))]

In [ ]:
#####################################

--Breakdown by Top vs Worst Sellers:

In [ ]:
#Find Top 10 Categories (with the addition of best sub-category within the category).

In [106]:
%%sql SELECT  Category, Sub_Category , round(cast(SUM(sales) as numeric),2) AS prod_sales
FROM store
GROUP BY Category,Sub_Category
ORDER BY prod_sales DESC;

 * postgresql://postgres:***@localhost:5432/sql_analysis
17 rows affected.


[('Technology', 'Phones', Decimal('330007.05')),
 ('Furniture', 'Chairs', Decimal('328449.10')),
 ('Office Supplies', 'Storage', Decimal('223843.61')),
 ('Furniture', 'Tables', Decimal('206965.53')),
 ('Office Supplies', 'Binders', Decimal('203401.19')),
 ('Technology', 'Machines', Decimal('189238.63')),
 ('Technology', 'Accessories', Decimal('167380.32')),
 ('Technology', 'Copiers', Decimal('149528.03')),
 ('Furniture', 'Bookcases', Decimal('114880.00')),
 ('Office Supplies', 'Appliances', Decimal('106152.99')),
 ('Furniture', 'Furnishings', Decimal('91705.16')),
 ('Office Supplies', 'Paper', Decimal('78479.21')),
 ('Office Supplies', 'Supplies', Decimal('46673.54')),
 ('Office Supplies', 'Art', Decimal('27118.79')),
 ('Office Supplies', 'Envelopes', Decimal('16476.40')),
 ('Office Supplies', 'Labels', Decimal('12486.31')),
 ('Office Supplies', 'Fasteners', Decimal('3024.28'))]

In [ ]:
#Find Top 10 Sub-Categories. :

In [107]:
%%sql SELECT round(cast(SUM(sales) as numeric),2) AS prod_sales,Sub_Category
FROM store
GROUP BY Sub_Category
ORDER BY prod_sales DESC
OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[(Decimal('328449.10'), 'Chairs'),
 (Decimal('223843.61'), 'Storage'),
 (Decimal('206965.53'), 'Tables'),
 (Decimal('203401.19'), 'Binders'),
 (Decimal('189238.63'), 'Machines'),
 (Decimal('167380.32'), 'Accessories'),
 (Decimal('149528.03'), 'Copiers'),
 (Decimal('114880.00'), 'Bookcases'),
 (Decimal('106152.99'), 'Appliances'),
 (Decimal('91705.16'), 'Furnishings')]

In [ ]:
#Find Worst 10 Categories.:

In [108]:
%%sql SELECT round(cast(SUM(sales) as numeric),2) AS prod_sales, Category, Sub_Category
FROM store
GROUP BY Category, Sub_Category
ORDER BY prod_sales;

 * postgresql://postgres:***@localhost:5432/sql_analysis
17 rows affected.


[(Decimal('3024.28'), 'Office Supplies', 'Fasteners'),
 (Decimal('12486.31'), 'Office Supplies', 'Labels'),
 (Decimal('16476.40'), 'Office Supplies', 'Envelopes'),
 (Decimal('27118.79'), 'Office Supplies', 'Art'),
 (Decimal('46673.54'), 'Office Supplies', 'Supplies'),
 (Decimal('78479.21'), 'Office Supplies', 'Paper'),
 (Decimal('91705.16'), 'Furniture', 'Furnishings'),
 (Decimal('106152.99'), 'Office Supplies', 'Appliances'),
 (Decimal('114880.00'), 'Furniture', 'Bookcases'),
 (Decimal('149528.03'), 'Technology', 'Copiers'),
 (Decimal('167380.32'), 'Technology', 'Accessories'),
 (Decimal('189238.63'), 'Technology', 'Machines'),
 (Decimal('203401.19'), 'Office Supplies', 'Binders'),
 (Decimal('206965.53'), 'Furniture', 'Tables'),
 (Decimal('223843.61'), 'Office Supplies', 'Storage'),
 (Decimal('328449.10'), 'Furniture', 'Chairs'),
 (Decimal('330007.05'), 'Technology', 'Phones')]

In [ ]:
#Find Worst 10 Sub-Categories. :

In [109]:
%%sql SELECT round(cast(SUM(sales) as numeric),2) AS prod_sales, sub_Category
FROM store
GROUP BY Sub_Category
ORDER BY prod_sales
OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[(Decimal('12486.31'), 'Labels'),
 (Decimal('16476.40'), 'Envelopes'),
 (Decimal('27118.79'), 'Art'),
 (Decimal('46673.54'), 'Supplies'),
 (Decimal('78479.21'), 'Paper'),
 (Decimal('91705.16'), 'Furnishings'),
 (Decimal('106152.99'), 'Appliances'),
 (Decimal('114880.00'), 'Bookcases'),
 (Decimal('149528.03'), 'Copiers'),
 (Decimal('167380.32'), 'Accessories')]

In [ ]:
#Show the Basic Order information.

In [110]:
%%sql select count(Order_ID) as Purchases,
round(cast(sum(Sales)as numeric),2) as Total_Sales,
round(cast(sum(((profit/((sales-profit))*100)))/ count(*)as numeric),2) as avg_percentage_profit,
min(Order_date) as first_purchase_date,
max(Order_date) as Latest_purchase_date,
count(distinct(Product_Name)) as Products_Purchased,
count(distinct(City)) as Location_count
from store

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[(9989, Decimal('2295810.15'), Decimal('32.38'), datetime.date(2013, 1, 3), datetime.date(2016, 12, 30), 1849, 531)]

In [ ]:
###########################

# RETURN LEVEL ANALYSIS

In [ ]:
#Find the number of returned orders.

In [111]:
%%sql select Returned_items, count(Returned_items)as Returned_Items_Count
from store
group by Returned_items
Having Returned_items='Returned'

 * postgresql://postgres:***@localhost:5432/sql_analysis
1 rows affected.


[('Returned', 800)]

In [ ]:
#Find Top 10 Returned Categories.:

In [112]:
%%sql SELECT Returned_items, Count(Returned_items) as no_of_returned ,Category, Sub_Category
FROM store
GROUP BY Returned_items,Category,Sub_Category
Having Returned_items='Returned'
ORDER BY count(Returned_items) DESC
limit 10;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('Returned', 137, 'Office Supplies', 'Binders'),
 ('Returned', 123, 'Office Supplies', 'Paper'),
 ('Returned', 77, 'Technology', 'Phones'),
 ('Returned', 72, 'Furniture', 'Furnishings'),
 ('Returned', 61, 'Technology', 'Accessories'),
 ('Returned', 57, 'Office Supplies', 'Storage'),
 ('Returned', 53, 'Furniture', 'Chairs'),
 ('Returned', 47, 'Office Supplies', 'Art'),
 ('Returned', 40, 'Office Supplies', 'Appliances'),
 ('Returned', 30, 'Furniture', 'Tables')]

In [ ]:
#Find Top 10  Returned Sub-Categories.:

In [113]:
%%sql SELECT Returned_items, Count(Returned_items),Sub_Category
FROM store
GROUP BY Returned_items, Sub_Category
Having Returned_items='Returned'
ORDER BY Count(Returned_items) DESC
OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('Returned', 123, 'Paper'),
 ('Returned', 77, 'Phones'),
 ('Returned', 72, 'Furnishings'),
 ('Returned', 61, 'Accessories'),
 ('Returned', 57, 'Storage'),
 ('Returned', 53, 'Chairs'),
 ('Returned', 47, 'Art'),
 ('Returned', 40, 'Appliances'),
 ('Returned', 30, 'Tables'),
 ('Returned', 21, 'Labels')]

In [ ]:
#Find Top 10 Customers Returned Frequently.:

In [114]:
%%sql SELECT Returned_items, Count(Returned_items) As Returned_Items_Count, Customer_ID,Customer_duration, States,City
FROM store
GROUP BY Returned_items,customer_name, Customer_ID,customer_duration,states,city
Having Returned_items='Returned'
ORDER BY Count(Returned_items) DESC
limit 10;

 * postgresql://postgres:***@localhost:5432/sql_analysis
10 rows affected.


[('Returned', 14, 'SV-20365                 ', 'old customer', 'New York', 'New York City'),
 ('Returned', 11, 'WB-21850                 ', 'old customer', 'California', 'Anaheim'),
 ('Returned', 10, 'CL-12565                 ', 'old customer', 'California', 'Los Angeles'),
 ('Returned', 9, 'SG-20080                 ', 'old customer', 'California', 'Redlands'),
 ('Returned', 8, 'SP-20620                 ', 'old customer', 'Pennsylvania', 'Philadelphia'),
 ('Returned', 8, 'DR-12880                 ', 'old customer', 'California', 'Inglewood'),
 ('Returned', 8, 'RP-19390                 ', 'old customer', 'Washington', 'Seattle'),
 ('Returned', 7, 'DH-13075                 ', 'old customer', 'California', 'Fairfield'),
 ('Returned', 7, 'DM-13015                 ', 'old customer', 'New York', 'New York City'),
 ('Returned', 7, 'RW-19540                 ', 'old customer', 'California', 'Brentwood')]

In [ ]:
#Find Top 20 cities and states having higher return.

In [115]:
%%sql SELECT Returned_items, Count(Returned_items)as Returned_Items_Count,States,City
FROM store
GROUP BY Returned_items,states,city
Having Returned_items='Returned'
ORDER BY Count(Returned_items) DESC
limit 20;

 * postgresql://postgres:***@localhost:5432/sql_analysis
20 rows affected.


[('Returned', 117, 'California', 'Los Angeles'),
 ('Returned', 64, 'California', 'San Francisco'),
 ('Returned', 60, 'Washington', 'Seattle'),
 ('Returned', 45, 'New York', 'New York City'),
 ('Returned', 18, 'California', 'San Diego'),
 ('Returned', 16, 'Arizona', 'Phoenix'),
 ('Returned', 13, 'Pennsylvania', 'Philadelphia'),
 ('Returned', 12, 'Illinois', 'Chicago'),
 ('Returned', 12, 'Oregon', 'Salem'),
 ('Returned', 12, 'California', 'Brentwood'),
 ('Returned', 11, 'California', 'Pasadena'),
 ('Returned', 11, 'California', 'Anaheim'),
 ('Returned', 10, 'California', 'Oakland'),
 ('Returned', 9, 'California', 'Fairfield'),
 ('Returned', 9, 'Texas', 'Houston'),
 ('Returned', 9, 'Georgia', 'Columbus'),
 ('Returned', 9, 'California', 'Redlands'),
 ('Returned', 8, 'Delaware', 'Newark'),
 ('Returned', 8, 'California', 'Inglewood'),
 ('Returned', 8, 'Colorado', 'Denver')]

In [ ]:
#Check whether new customers are returning higher or not.

In [116]:
%%sql SELECT Returned_items, Count(Returned_items)as Returned_Items_Count,Customer_duration
FROM store
GROUP BY Returned_items,Customer_duration
Having Returned_items='Returned'
ORDER BY Count(Returned_items) DESC
limit 20;

 * postgresql://postgres:***@localhost:5432/sql_analysis
2 rows affected.


[('Returned', 796, 'old customer'), ('Returned', 4, 'new customer')]

In [ ]:
#Find Top  Reasons for returning.

In [117]:
%%sql SELECT Returned_items, Count(Returned_items)as Returned_Items_Count,return_reason
FROM store
GROUP BY Returned_items,return_reason
Having Returned_items='Returned'
ORDER BY Count(Returned_items) DESC

 * postgresql://postgres:***@localhost:5432/sql_analysis
6 rows affected.


[('Returned', 304, 'Product Description Inaccurate'),
 ('Returned', 158, 'Customer Dissatified With Product'),
 ('Returned', 100, 'Product Arrived After Promised Delivery Date'),
 ('Returned', 90, 'Product Damaged'),
 ('Returned', 81, 'Product Ordered Accidently'),
 ('Returned', 67, 'Incorrect Products Delivered')]